cirq-core/cirq/circuits/moment.py

```diff
@@ -168,6 +168,11 @@ def with_operation(self, operation: 'cirq.Operation') -> 'cirq.Moment':
         m._operations = self._operations + (operation,)
         m._qubits = frozenset(self._qubits.union(set(operation.qubits)))
         m._qubit_to_op = self._qubit_to_op.copy()
+
+        m._measurement_key_objs = set(self._measurement_key_objs or {})
+        m._control_keys = set(self._control_keys or {})
+        m._measurement_key_objs |= protocols.measurement_key_objs(operation)
+        m._control_keys |= protocols.control_keys(operation)
```

Comment:
These will be incorrect if `self._measurement_key_objs` or `self._control_keys` have not yet been computed on _this_ moment. Instead, should access these through the methods that will populate them if needed:
```suggestion
        m._measurement_key_objs = set(self._measurement_key_objs_())
        m._measurement_key_objs |= protocols.measurement_key_objs(operation)
        m._control_keys = set(self._control_keys_())
        m._control_keys |= protocols.control_keys(operation)
```
Also, it strikes me that maybe this should just delegate to `self.with_operations(operation)` to avoid duplicating most of the code. I suspect the cost of the loop over one operation would be negligible.

Reply:
oh doh, I went through like 5 iterations of this thing trying to see if I could get more speed and ended up with the bad one

The problem with delegating is that `self.with_operations` creates a new Moment while copying over all the information each time, so you are constantly creating and destroying a bunch of sets which I think would be bad.

Reply:
fixed